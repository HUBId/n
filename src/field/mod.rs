//! Field arithmetic primitives for the `rpp-stark` proof system.
//! Contains finite field implementations and polynomial utilities.

pub mod polynomial;
pub mod prime_field;

pub use prime_field::FieldElement;

#[cfg(test)]
pub mod tests;
