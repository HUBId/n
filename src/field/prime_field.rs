//! Prime field interface tailored for the `rpp-stark` engine.
//!
//! This module intentionally exposes *only* the type- and trait-level
//! contracts that downstream back-ends must satisfy.  No arithmetic logic
//! is provided here; consumers are expected to supply constant-time
//! implementations that respect the documented invariants.
//!
//! # Security Notes
//!
//! * No `unsafe` code is used in this module.
//! * Randomness must always be derived deterministically from the transcript;
//!   operating-system entropy sources are intentionally avoided.

use core::{cmp::Ordering, fmt};

/// Metadata describing the underlying field modulus.
#[derive(Debug, Clone, Copy)]
pub struct Modulus {
    /// Prime modulus value in canonical representation.
    pub value: u64,
    /// Indicates whether the modulus passed primality checks during configuration.
    pub is_prime: bool,
}

impl Modulus {
    /// Creates a new modulus descriptor.
    pub const fn new(value: u64, is_prime: bool) -> Self {
        Self { value, is_prime }
    }
}

/// Canonical generator for the default field used across the system.
pub const DEFAULT_MODULUS: Modulus = Modulus::new(0xffffffff00000001, true);

/// Field element represented as a canonical value modulo the prime.
///
/// # Representation
///
/// * `FieldElement` is a transparent wrapper around a raw `u64`.  When the
///   element is in canonical (non-Montgomery) form the wrapped integer must be
///   within the range `[0, MODULUS.value)`.
/// * Montgomery form encodes elements as `a * R mod MODULUS`, where `R = 2^64`.
/// * Serialization uses **little-endian** byte order for canonical
///   representations; Montgomery encodings are documented separately by the
///   conversion traits.
///
/// # Safety & Auditing
///
/// The type itself performs no runtime validation.  Implementations of the
/// provided traits must enforce the documented invariants and perform
/// constant-time assertions where required by `ConstantTimeAssertions`.
#[repr(transparent)]
#[derive(Clone, Copy, PartialEq, Eq, Default)]
pub struct FieldElement(pub u64);

impl fmt::Debug for FieldElement {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_tuple("FieldElement").field(&self.0).finish()
    }
}

impl FieldElement {
    /// Canonical prime modulus associated with this field.
    pub const MODULUS: Modulus = DEFAULT_MODULUS;
    /// Montgomery radix `R = 2^64 mod MODULUS`.
    pub const R: u64 = 0xffffffff;
    /// Precomputed `R^2 mod MODULUS` used for Montgomery conversions.
    pub const R2: u64 = 0xfffffffe00000001;
    /// `-MODULUS^{-1} mod 2^64`, the Montgomery reduction parameter.
    pub const MONTGOMERY_INV: u64 = 0xfffffffeffffffff;
    /// Designated generator for the multiplicative subgroup.
    pub const GENERATOR: FieldElement = FieldElement(3);
    /// Additive identity in canonical form.
    pub const ZERO: FieldElement = FieldElement(0);
    /// Multiplicative identity in canonical form.
    pub const ONE: FieldElement = FieldElement(1);

    #[inline(always)]
    const fn modulus() -> u64 {
        Self::MODULUS.value
    }

    #[inline(always)]
    fn reduce_once(value: u64) -> u64 {
        let reduced = value.wrapping_sub(Self::modulus());
        let underflow = (reduced > value) as u64;
        let mask = underflow.wrapping_neg();
        reduced.wrapping_add(Self::modulus() & mask)
    }

    #[inline(always)]
    fn montgomery_reduce(t: u128) -> u64 {
        let modulus = Self::modulus() as u128;
        let m = (t as u64).wrapping_mul(Self::MONTGOMERY_INV) as u128;
        let u = t.wrapping_add(m.wrapping_mul(modulus)) >> 64;
        let candidate = u as u64;
        Self::reduce_once(candidate)
    }

    #[inline(always)]
    fn montgomery_mul(a: u64, b: u64) -> u64 {
        let product = (a as u128) * (b as u128);
        Self::montgomery_reduce(product)
    }

    #[inline(always)]
    fn add_internal(&self, rhs: &Self) -> Self {
        let modulus = Self::modulus() as u128;
        let sum = self.0 as u128 + rhs.0 as u128;
        let mask = (sum >= modulus) as u128;
        let result = sum - mask * modulus;
        FieldElement(result as u64)
    }

    #[inline(always)]
    fn sub_internal(&self, rhs: &Self) -> Self {
        let (diff, borrow) = self.0.overflowing_sub(rhs.0);
        let mask = (borrow as u64).wrapping_neg();
        let result = diff.wrapping_add(Self::modulus() & mask);
        FieldElement(result)
    }

    #[inline(always)]
    fn neg_internal(&self) -> Self {
        let tmp = Self::modulus().wrapping_sub(self.0);
        let nz = (self.0 != 0) as u64;
        let mask = nz.wrapping_neg();
        FieldElement(tmp & mask)
    }

    #[inline(always)]
    fn mul_internal(&self, rhs: &Self) -> Self {
        let modulus = Self::modulus() as u128;
        let product = (self.0 as u128 * rhs.0 as u128) % modulus;
        FieldElement(product as u64)
    }

    #[inline(always)]
    fn square_internal(&self) -> Self {
        let modulus = Self::modulus() as u128;
        let product = (self.0 as u128 * self.0 as u128) % modulus;
        FieldElement(product as u64)
    }

    /// Raises the element to the provided exponent using square-and-multiply.
    pub fn pow(&self, exponent: u64) -> Self {
        let mut base = *self;
        let mut acc = FieldElement::ONE;
        for i in 0..64 {
            let bit = (exponent >> i) & 1;
            let mask = (bit as u64).wrapping_neg();
            let candidate = acc.mul_internal(&base);
            let acc_val = (acc.0 & !mask) | (candidate.0 & mask);
            acc = FieldElement(acc_val);
            base = base.square_internal();
        }
        acc
    }

    /// Converts uniformly distributed transcript bytes into a field element.
    pub fn from_transcript_bytes(bytes: &[u8; 32]) -> Self {
        let mut limbs = [0u64; 4];
        for (dst, chunk) in limbs.iter_mut().zip(bytes.chunks_exact(8)) {
            let mut buf = [0u8; 8];
            buf.copy_from_slice(chunk);
            *dst = u64::from_le_bytes(buf);
        }

        let mut acc = FieldElement::ZERO;
        let mut factor = FieldElement::ONE;
        let radix = FieldElement::from(Self::R);
        for limb in limbs.iter() {
            let term = FieldElement::from(*limb).mul_internal(&factor);
            acc = acc.add_internal(&term);
            factor = factor.mul_internal(&radix);
        }
        acc
    }
}

/// Trait describing the high-level arithmetic contract for field elements.
pub trait FieldElementOps: Sized {
    /// Adds two canonical field elements, returning the canonical representative.
    fn add(&self, rhs: &Self) -> Self;
    /// Subtracts `rhs` from `self` in canonical form.
    fn sub(&self, rhs: &Self) -> Self;
    /// Computes the additive inverse of `self`.
    fn neg(&self) -> Self;
    /// Multiplies two field elements.
    fn mul(&self, rhs: &Self) -> Self;
    /// Squares the field element.
    fn square(&self) -> Self;
    /// Computes the multiplicative inverse, returning `None` for zero.
    fn inv(&self) -> Option<Self>;
}

/// Trait capturing Montgomery encoding/decoding contracts.
pub trait MontgomeryConvertible: Sized {
    /// Converts a canonical element into Montgomery form.
    fn to_montgomery(&self) -> Self;
    /// Converts an element from Montgomery form back to canonical representation.
    fn from_montgomery(&self) -> Self;
}

/// Trait defining serialization requirements for field elements.
pub trait CanonicalSerialize: Sized {
    /// Canonical serialization output type (e.g. `[u8; 8]`).
    type Bytes;

    /// Serializes the element into canonical little-endian bytes.
    fn to_bytes(&self) -> Self::Bytes;

    /// Attempts to deserialize from canonical little-endian bytes.
    fn from_bytes(bytes: &Self::Bytes) -> Result<Self, FieldDeserializeError>;
}

/// Errors that can occur while deserializing field elements from canonical bytes.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum FieldDeserializeError {
    /// The provided bytes encode a value outside of the canonical field range.
    FieldDeserializeNonCanonical,
}

impl fmt::Display for FieldDeserializeError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            FieldDeserializeError::FieldDeserializeNonCanonical => {
                f.write_str("field element deserialization failed: non-canonical input")
            }
        }
    }
}

/// Trait collecting constant-time auditing assertions required by higher layers.
pub trait ConstantTimeAssertions {
    /// Ensures the element is strictly less than the field modulus in constant time.
    fn assert_lt_modulus(&self);
    /// Ensures the element is *not* the additive identity in constant time.
    fn assert_nonzero(&self);
    /// Ensures the element is a valid canonical representative (or panics/logs otherwise).
    fn assert_canonical(&self);
}

impl FieldElementOps for FieldElement {
    fn add(&self, rhs: &Self) -> Self {
        self.add_internal(rhs)
    }

    fn sub(&self, rhs: &Self) -> Self {
        self.sub_internal(rhs)
    }

    fn neg(&self) -> Self {
        self.neg_internal()
    }

    fn mul(&self, rhs: &Self) -> Self {
        self.mul_internal(rhs)
    }

    fn square(&self) -> Self {
        self.square_internal()
    }

    fn inv(&self) -> Option<Self> {
        if self.0 == 0 {
            None
        } else {
            Some(self.pow(Self::modulus() - 2))
        }
    }
}

impl MontgomeryConvertible for FieldElement {
    fn to_montgomery(&self) -> Self {
        FieldElement(Self::montgomery_mul(self.0, Self::R2))
    }

    fn from_montgomery(&self) -> Self {
        FieldElement(Self::montgomery_mul(self.0, 1))
    }
}

impl CanonicalSerialize for FieldElement {
    type Bytes = [u8; 8];

    fn to_bytes(&self) -> Self::Bytes {
        self.assert_canonical();
        self.0.to_le_bytes()
    }

    fn from_bytes(bytes: &Self::Bytes) -> Result<Self, FieldDeserializeError> {
        let value = u64::from_le_bytes(*bytes);
        if value < Self::modulus() {
            Ok(FieldElement(value))
        } else {
            Err(FieldDeserializeError::FieldDeserializeNonCanonical)
        }
    }
}

impl ConstantTimeAssertions for FieldElement {
    fn assert_lt_modulus(&self) {
        if self.0 >= Self::modulus() {
            panic!("field element not canonical");
        }
    }

    fn assert_nonzero(&self) {
        if self.0 == 0 {
            panic!("field element is zero");
        }
    }

    fn assert_canonical(&self) {
        self.assert_lt_modulus();
    }
}

impl From<u64> for FieldElement {
    fn from(value: u64) -> Self {
        FieldElement(FieldElement::reduce_once(value))
    }
}

impl From<FieldElement> for u64 {
    fn from(value: FieldElement) -> Self {
        value.assert_canonical();
        value.0
    }
}

impl Ord for FieldElement {
    fn cmp(&self, other: &Self) -> Ordering {
        self.0.cmp(&other.0)
    }
}

impl PartialOrd for FieldElement {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl FieldElement {
    /// Returns `true` if the element is the additive identity.
    pub fn is_zero(&self) -> bool {
        self.0 == 0
    }

    /// Returns `true` if the element is the multiplicative identity.
    pub fn is_one(&self) -> bool {
        self.0 == 1
    }
}
