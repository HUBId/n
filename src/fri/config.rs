//! Canonical configuration descriptions for the FRI protocol.
//! Provides named profiles that are shared between prover and verifier.

/// Canonical digest representation for a FRI parameter set.
///
/// The digest is computed over the canonical serialization of the
/// folding factor, folding mode, query schedule, and target depth
/// using BLAKE3.
/// The hexadecimal strings exposed below allow integrators to pin
/// protocol parameters without embedding executable verification logic.
pub type ParameterDigest = &'static str;

/// Describes a complete FRI profile that can be referenced by the
/// STARK configuration and transcript builders.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct FriProfile {
    /// Human readable profile identifier.
    pub name: &'static str,
    /// Folding factor applied to each layer; we expose the canonical binary
    /// folding schedule in this module.
    pub folding_factor: usize,
    /// Number of sampling queries performed by the verifier.
    pub query_count: usize,
    /// Target depth (logarithmic size) of the final layer after folding.
    pub target_depth: usize,
    /// Digest binding the profile to a concrete transcript layout.
    pub parameter_digest: ParameterDigest,
}

/// Standard security profile targeting typical rollup deployments.
pub const STANDARD_FRI_PROFILE: FriProfile = FriProfile {
    name: "standard",
    folding_factor: 2,
    query_count: 64,
    target_depth: 6,
    parameter_digest: "5570706d57436b4fa198d0e8c8a6d3fa01d8c348c602248b066ee8f3fe0eb566",
};

/// High security profile with an increased query budget.
pub const HISEC_FRI_PROFILE: FriProfile = FriProfile {
    name: "hisec",
    folding_factor: 2,
    query_count: 96,
    target_depth: 8,
    parameter_digest: "40b923845a41a78da7bdef951f541ae1b6b30794b0244a54b19fc5801d686b06",
};
