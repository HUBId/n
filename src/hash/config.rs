//! Hash configuration definitions shared across prover and verifier.
//! Parameters ensure deterministic hash usage within transcripts and commitments.

/// Poseidon parameter set descriptor.
#[derive(Debug, Clone)]
pub struct PoseidonParameters {
    /// Number of full rounds in the permutation.
    pub full_rounds: usize,
    /// Number of partial rounds in the permutation.
    pub partial_rounds: usize,
    /// Width of the permutation state.
    pub width: usize,
}

impl PoseidonParameters {
    /// Creates a new parameter set with deterministic defaults.
    pub fn new(full_rounds: usize, partial_rounds: usize, width: usize) -> Self {
        Self {
            full_rounds,
            partial_rounds,
            width,
        }
    }
}

/// Blake3 parameter descriptor for transcript usage.
#[derive(Debug, Clone)]
pub struct Blake3Parameters {
    /// Fixed domain separation label.
    pub label: &'static [u8],
}

impl Blake3Parameters {
    /// Creates a new descriptor.
    pub const fn new(label: &'static [u8]) -> Self {
        Self { label }
    }
}

/// Aggregate hash parameters used by the STARK engine.
#[derive(Debug, Clone)]
pub struct HashParameters {
    /// Poseidon permutation configuration.
    pub poseidon: PoseidonParameters,
    /// Blake3 transcript configuration.
    pub blake3: Blake3Parameters,
}

impl HashParameters {
    /// Constructs aggregate parameters.
    pub fn new(poseidon: PoseidonParameters, blake3: Blake3Parameters) -> Self {
        Self { poseidon, blake3 }
    }
}
