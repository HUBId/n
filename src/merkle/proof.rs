use crate::params::{Endianness, MerkleArity, StarkParams};

use super::traits::MerkleHasher;
use super::tree::CommitAux;
use super::types::{Digest, Leaf, MerkleArityExt, MerkleError, ProofNode};

use crate::proof::types::PROOF_VERSION;

/// Canonical Merkle opening containing a batch of indices.
#[derive(Clone, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct MerkleProof {
    pub version: u16,
    pub arity: MerkleArity,
    pub leaf_encoding: Endianness,
    pub path: Vec<ProofNode>,
    pub indices: Vec<u32>,
    pub leaf_width: u8,
    pub domain_sep: u64,
    pub leaf_width_bytes: u32,
    pub digest_size: u16,
}

impl MerkleProof {
    pub fn path(&self) -> &[ProofNode] {
        &self.path
    }
}

/// Deterministic builder for Merkle openings.
pub struct ProofBuilder<'a> {
    aux: &'a CommitAux,
}

impl<'a> ProofBuilder<'a> {
    pub fn new(aux: &'a CommitAux) -> Self {
        Self { aux }
    }

    pub fn open(&self, params: &StarkParams, indices: &[u32]) -> Result<MerkleProof, MerkleError> {
        self.aux.ensure_openable(params, PROOF_VERSION)?;
        if indices.is_empty() {
            return Err(MerkleError::InvalidPathLength);
        }
        let mut sorted = indices.to_vec();
        sorted.sort_unstable();
        for window in sorted.windows(2) {
            if window[0] == window[1] {
                return Err(MerkleError::DuplicateIndex { index: window[1] });
            }
        }
        let leaves = self.aux.leaves();
        for &index in &sorted {
            if index as usize >= leaves {
                return Err(MerkleError::IndexOutOfRange {
                    index,
                    max: leaves.saturating_sub(1) as u32,
                });
            }
        }

        let arity = self.aux.arity().as_usize();
        let mut current = sorted.clone();
        let mut path = Vec::new();

        for depth in 0..self.aux.height().saturating_sub(1) {
            let level = self.aux.level(depth).unwrap_or(&[]);
            let mut next = Vec::new();
            let mut i = 0;
            while i < current.len() {
                let index = current[i];
                let parent = index / arity as u32;
                let mut group_end = i + 1;
                while group_end < current.len() && current[group_end] / arity as u32 == parent {
                    group_end += 1;
                }
                let mut present = vec![false; arity];
                for child in &current[i..group_end] {
                    present[(child % arity as u32) as usize] = true;
                }
                let missing: Vec<usize> = (0..arity).filter(|pos| !present[*pos]).collect();
                if !missing.is_empty() {
                    let mut node = new_proof_node(self.aux.arity(), self.aux.digest_size());
                    for (slot, position) in missing.iter().enumerate() {
                        let absolute = parent as usize * arity + position;
                        let digest = if absolute < level.len() {
                            level[absolute].clone()
                        } else {
                            level
                                .last()
                                .cloned()
                                .unwrap_or_else(|| Digest::zero(self.aux.digest_size()))
                        };
                        node.siblings_mut()[slot] = digest;
                    }
                    path.push(node);
                }
                next.push(parent);
                i = group_end;
            }
            current = next;
        }

        Ok(MerkleProof {
            version: PROOF_VERSION,
            arity: self.aux.arity(),
            leaf_encoding: self.aux.leaf_encoding,
            path,
            indices: sorted,
            leaf_width: self.aux.leaf_width,
            domain_sep: self.aux.domain_sep(),
            leaf_width_bytes: (self.aux.leaf_width as u32)
                * super::tree::field_element_size(params.field()) as u32,
            digest_size: self.aux.digest_size() as u16,
        })
    }
}

fn new_proof_node(arity: MerkleArity, digest_size: usize) -> ProofNode {
    match arity {
        MerkleArity::Binary => {
            ProofNode::Arity2(core::array::from_fn(|_| Digest::zero(digest_size)))
        }
        MerkleArity::Quaternary => {
            ProofNode::Arity4(core::array::from_fn(|_| Digest::zero(digest_size)))
        }
    }
}

fn ensure_indices_sorted(indices: &[u32]) -> Result<(), MerkleError> {
    for window in indices.windows(2) {
        if window[0] >= window[1] {
            return Err(MerkleError::DuplicateIndex { index: window[1] });
        }
    }
    Ok(())
}

pub fn verify_proof<H: MerkleHasher>(
    params: &StarkParams,
    root: &Digest,
    proof: &MerkleProof,
    leaves: &[Leaf],
) -> Result<(), MerkleError> {
    if proof.version != PROOF_VERSION {
        return Err(MerkleError::ProofVersionMismatch {
            expected: PROOF_VERSION,
            got: proof.version,
        });
    }
    if proof.arity != params.merkle().arity {
        return Err(MerkleError::ArityMismatch);
    }
    if proof.leaf_encoding != params.merkle().leaf_encoding {
        return Err(MerkleError::IncompatibleParams {
            reason: "leaf encoding mismatch",
        });
    }
    if proof.leaf_width != params.merkle().leaf_width {
        return Err(MerkleError::LeafWidthMismatch {
            expected: params.merkle().leaf_width,
            got: proof.leaf_width,
        });
    }
    if proof.domain_sep != params.merkle().domain_sep {
        return Err(MerkleError::IncompatibleParams {
            reason: "domain separation mismatch",
        });
    }
    if proof.indices.len() != leaves.len() {
        return Err(MerkleError::InvalidPathLength);
    }
    if proof.digest_size as usize != H::digest_size() {
        return Err(MerkleError::IncompatibleParams {
            reason: "digest size mismatch",
        });
    }
    ensure_indices_sorted(&proof.indices)?;

    let element_size = super::tree::field_element_size(params.field());
    let expected = element_size * proof.leaf_width as usize;
    let mut current: Vec<(u32, H::Digest)> = Vec::with_capacity(leaves.len());
    for (&index, leaf) in proof.indices.iter().zip(leaves.iter()) {
        if leaf.as_bytes().len() != expected {
            let got = (leaf.as_bytes().len() / element_size) as u8;
            return Err(MerkleError::LeafWidthMismatch {
                expected: proof.leaf_width,
                got,
            });
        }
        let digest = H::hash_leaves(proof.domain_sep, leaf.as_bytes());
        current.push((index, digest));
    }

    if current.is_empty() {
        return Err(MerkleError::InvalidPathLength);
    }

    let mut path_iter = proof.path.iter();
    let arity = proof.arity.as_usize() as u32;

    while !(current.len() == 1 && path_iter.clone().next().is_none()) {
        if current.is_empty() {
            return Err(MerkleError::InvalidPathLength);
        }
        let mut next = Vec::new();
        let mut i = 0;
        while i < current.len() {
            let (index, _) = current[i];
            let parent = index / arity;
            let mut group_end = i + 1;
            while group_end < current.len() && current[group_end].0 / arity == parent {
                group_end += 1;
            }
            let mut children: Vec<Option<H::Digest>> = vec![None; arity as usize];
            for &(child_index, digest) in &current[i..group_end] {
                let position = (child_index % arity) as usize;
                children[position] = Some(digest);
            }
            let missing = children.iter().filter(|child| child.is_none()).count();
            if missing > 0 {
                let node = path_iter.next().ok_or(MerkleError::InvalidPathLength)?;
                let siblings = node.siblings();
                if siblings.len() != (arity as usize - 1) {
                    return Err(MerkleError::InvalidPathLength);
                }
                let mut sibling_iter = siblings.iter();
                for child in children.iter_mut() {
                    if child.is_none() {
                        if let Some(raw) = sibling_iter.next() {
                            let digest = H::from_bytes(raw.as_bytes())
                                .ok_or(MerkleError::VerificationFailed)?;
                            *child = Some(digest);
                        }
                    }
                }
                for raw in sibling_iter {
                    if raw.as_bytes().iter().any(|byte| *byte != 0) {
                        return Err(MerkleError::InvalidPathLength);
                    }
                }
            }

            let mut ordered = Vec::with_capacity(arity as usize);
            for child in children {
                let digest = child.ok_or(MerkleError::VerificationFailed)?;
                ordered.push(digest);
            }
            let parent_digest = H::hash_nodes(proof.domain_sep, &ordered);
            next.push((parent, parent_digest));
            i = group_end;
        }
        next.sort_by_key(|entry| entry.0);
        current = next;
    }

    let computed = current.first().ok_or(MerkleError::InvalidPathLength)?;
    let root_digest = H::from_bytes(root.as_bytes()).ok_or(MerkleError::VerificationFailed)?;
    if computed.1.as_ref() == root_digest.as_ref() {
        Ok(())
    } else {
        Err(MerkleError::VerificationFailed)
    }
}
