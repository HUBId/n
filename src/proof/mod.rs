//! Proof generation and verification subsystem.
//! Provides embedded prover and verifier implementations for the STARK engine.

pub mod prover;
pub mod verifier;

pub use prover::Prover;
pub use verifier::Verifier;
