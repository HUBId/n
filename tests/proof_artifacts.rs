use insta::assert_json_snapshot;
use rpp_stark::config::{
    build_proof_system_config, build_prover_context, compute_param_digest, ChunkingPolicy,
    CommonIdentifiers, ProfileConfig, ProofSystemConfig, ProverContext, ThreadPoolProfile,
    COMMON_IDENTIFIERS, PROFILE_STANDARD_CONFIG,
};
use rpp_stark::field::prime_field::{CanonicalSerialize, FieldElementOps};
use rpp_stark::field::FieldElement;
use rpp_stark::proof::public_inputs::{
    ExecutionHeaderV1, ProofKind, PublicInputVersion, PublicInputs,
};
use rpp_stark::utils::serialization::{ProofBytes, WitnessBlob};
use std::collections::BTreeMap;

fn hex<const N: usize>(bytes: &[u8; N]) -> String {
    bytes.iter().map(|byte| format!("{:02x}", byte)).collect()
}

fn make_public_inputs<'a>(header: &'a ExecutionHeaderV1, body: &'a [u8]) -> PublicInputs<'a> {
    PublicInputs::Execution {
        header: header.clone(),
        body,
    }
}

struct SnapshotSetup {
    config: ProofSystemConfig,
    prover_context: ProverContext,
    header: ExecutionHeaderV1,
    body: Vec<u8>,
    witness: Vec<u8>,
}

impl SnapshotSetup {
    fn new() -> Self {
        let profile: ProfileConfig = PROFILE_STANDARD_CONFIG.clone();
        let common: CommonIdentifiers = COMMON_IDENTIFIERS.clone();
        let param_digest = compute_param_digest(&profile, &common);
        let config = build_proof_system_config(&profile, &param_digest);
        let prover_context = build_prover_context(
            &profile,
            &common,
            &param_digest,
            ThreadPoolProfile::SingleThread,
            ChunkingPolicy {
                min_chunk_items: 4,
                max_chunk_items: 32,
                stride: 1,
            },
        );

        let seed = FieldElement::from(3u64);
        let length = 128usize;
        let header = ExecutionHeaderV1 {
            version: PublicInputVersion::V1,
            program_digest: rpp_stark::utils::serialization::DigestBytes { bytes: [0u8; 32] },
            trace_length: length as u32,
            trace_width: 1,
        };
        let body = seed
            .to_bytes()
            .expect("fixture seed must be canonical")
            .to_vec();
        let witness = build_witness(seed, length);

        Self {
            config,
            prover_context,
            header,
            body,
            witness,
        }
    }
}

const LFSR_ALPHA: u64 = 5;
const LFSR_BETA: u64 = 7;

fn build_witness(seed: FieldElement, rows: usize) -> Vec<u8> {
    let alpha = FieldElement::from(LFSR_ALPHA);
    let beta = FieldElement::from(LFSR_BETA);
    let mut column = Vec::with_capacity(rows);
    let mut state = seed;
    column.push(state);
    for _ in 1..rows {
        state = state.mul(&alpha).add(&beta);
        column.push(state);
    }

    let mut bytes = Vec::with_capacity(20 + rows * 8);
    bytes.extend_from_slice(&(rows as u32).to_le_bytes());
    bytes.extend_from_slice(&1u32.to_le_bytes());
    bytes.extend_from_slice(&0u32.to_le_bytes());
    bytes.extend_from_slice(&0u32.to_le_bytes());
    bytes.extend_from_slice(&0u32.to_le_bytes());
    for value in column {
        let encoded = value.to_bytes().expect("fixture values must be canonical");
        bytes.extend_from_slice(&encoded);
    }
    bytes
}

fn decode_proof(bytes: &ProofBytes) -> rpp_stark::Proof {
    rpp_stark::Proof::from_bytes(bytes.as_slice()).expect("decode proof")
}

#[test]
fn snapshot_execution_proof_artifacts() {
    let setup = SnapshotSetup::new();
    let witness = WitnessBlob {
        bytes: &setup.witness,
    };
    let public_inputs = make_public_inputs(&setup.header, &setup.body);

    let proof_bytes = rpp_stark::generate_proof(
        ProofKind::Execution,
        &public_inputs,
        witness,
        &setup.config,
        &setup.prover_context,
    )
    .expect("proof generation succeeds");
    let decoded = decode_proof(&proof_bytes);

    let trace_paths: Vec<usize> = decoded
        .openings
        .trace
        .paths
        .iter()
        .map(|path| path.nodes.len())
        .collect();
    let composition_paths: Option<Vec<usize>> = decoded
        .openings
        .composition
        .as_ref()
        .map(|comp| comp.paths.iter().map(|path| path.nodes.len()).collect());
    let trace_indices = {
        let mut indices = decoded.openings.trace.indices.clone();
        indices.sort_unstable();
        indices
    };
    let composition_indices = decoded.openings.composition.as_ref().map(|comp| {
        let mut indices = comp.indices.clone();
        indices.sort_unstable();
        indices
    });
    let fri_positions: Vec<usize> = decoded
        .fri_proof
        .queries
        .iter()
        .map(|query| query.position)
        .collect();
    let fri_layer_path_lengths: Vec<usize> = decoded
        .fri_proof
        .queries
        .iter()
        .flat_map(|query| query.layers.iter().map(|layer| layer.path.len()))
        .collect();

    let trace_path_summary = summarize_lengths(&trace_paths);
    let composition_path_summary = composition_paths
        .as_ref()
        .map(|paths| summarize_lengths(paths))
        .unwrap_or_else(|| serde_json::json!({"total": 0, "histogram": {}}));
    let fri_path_summary = summarize_lengths(&fri_layer_path_lengths);

    let artifact = serde_json::json!({
        "trace_root": hex(&decoded.merkle.core_root),
        "composition_root": hex(&decoded.merkle.aux_root),
        "fri_roots": decoded.merkle.fri_layer_roots.iter().map(hex).collect::<Vec<_>>(),
        "trace_query_indices": trace_indices,
        "composition_query_indices": composition_indices,
        "fri_query_positions": fri_positions,
        "trace_path_lengths": trace_path_summary,
        "composition_path_lengths": composition_path_summary,
        "fri_path_lengths": fri_path_summary,
    });

    assert_json_snapshot!("execution_proof_artifacts", artifact);
}

fn summarize_lengths(lengths: &[usize]) -> serde_json::Value {
    let mut histogram: BTreeMap<usize, usize> = BTreeMap::new();
    for &len in lengths {
        *histogram.entry(len).or_insert(0) += 1;
    }
    serde_json::json!({
        "total": lengths.len(),
        "histogram": histogram,
    })
}
