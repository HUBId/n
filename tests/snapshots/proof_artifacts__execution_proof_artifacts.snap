---
source: tests/proof_artifacts.rs
expression: artifact
---
{
  "composition_path_lengths": {
    "histogram": {
      "10": 64
    },
    "total": 64
  },
  "composition_query_indices": [
    3,
    26,
    42,
    45,
    59,
    74,
    117,
    118,
    127,
    140,
    172,
    174,
    191,
    197,
    213,
    214,
    234,
    243,
    266,
    276,
    324,
    344,
    371,
    373,
    377,
    384,
    397,
    412,
    433,
    455,
    471,
    479,
    492,
    522,
    555,
    572,
    577,
    584,
    592,
    604,
    621,
    670,
    671,
    696,
    700,
    729,
    734,
    739,
    745,
    746,
    757,
    772,
    813,
    819,
    822,
    825,
    840,
    900,
    914,
    956,
    977,
    1015,
    1019,
    1021
  ],
  "composition_root": "b00162d607f98f5fc4d3eda621b26caf7ad62fcbe7993b43cd4bdb6324f5f725",
  "fri_path_lengths": {
    "histogram": {
      "10": 64,
      "6": 64,
      "7": 64,
      "8": 64,
      "9": 64
    },
    "total": 320
  },
  "fri_query_positions": [
    3,
    26,
    42,
    45,
    59,
    74,
    117,
    118,
    127,
    140,
    172,
    174,
    191,
    197,
    213,
    214,
    234,
    243,
    266,
    276,
    324,
    344,
    371,
    373,
    377,
    384,
    397,
    412,
    433,
    455,
    471,
    479,
    492,
    522,
    555,
    572,
    577,
    584,
    592,
    604,
    621,
    670,
    671,
    696,
    700,
    729,
    734,
    739,
    745,
    746,
    757,
    772,
    813,
    819,
    822,
    825,
    840,
    900,
    914,
    956,
    977,
    1015,
    1019,
    1021
  ],
  "fri_roots": [
    "251602d1ac89355a0bae915d3e3119528228bcd210fd04b3c5f639f28e1c0a66",
    "0323f8d5138650be0fa834d9e3793d9523e6eac97ad9ceee1b063feecd435c0f",
    "b2defab4ee1faf9b2925566afdb6a3c31a39a60229cd28539f0ae43081105d8f",
    "41dce5909d545bb0e2bf4c04efce1c99d31f603033348d0d9c00b702ec33b946",
    "3f592bd94f0a0a65f3eb304f08f83231f55932d9e1e9becb8ad5e0fc952830e8"
  ],
  "trace_path_lengths": {
    "histogram": {
      "10": 64
    },
    "total": 64
  },
  "trace_query_indices": [
    3,
    26,
    42,
    45,
    59,
    74,
    117,
    118,
    127,
    140,
    172,
    174,
    191,
    197,
    213,
    214,
    234,
    243,
    266,
    276,
    324,
    344,
    371,
    373,
    377,
    384,
    397,
    412,
    433,
    455,
    471,
    479,
    492,
    522,
    555,
    572,
    577,
    584,
    592,
    604,
    621,
    670,
    671,
    696,
    700,
    729,
    734,
    739,
    745,
    746,
    757,
    772,
    813,
    819,
    822,
    825,
    840,
    900,
    914,
    956,
    977,
    1015,
    1019,
    1021
  ],
  "trace_root": "2edc2d4e2384e2055871c3f6f3f202a585be09af46ed09a03c9550f27dd306d7"
}
